modules/effects/src/main/scala/dev/profunktor/redis4cats/transactions.scala

```diff
@@ -60,4 +60,94 @@ object transactions {
       }
   }
 
+  sealed trait RedisTransactionB[F[_], K, V, A] {
+    def transact(commands: RedisCommands[F, K, V])(implicit F: Concurrent[F], L: Log[F]): F[A]
+
+  }
+  object RedisTransactionB {
+
+    def builder[F[_], K, V]: Builder[F, K, V] = new Builder[F, K, V]
+
+    /**
+      * The Builder is a Partially Applied Constructor Allowing operations to fully infer.
+     **/
+    class Builder[F[_], K, V] {
+      def operation[A](r: RedisCommands[F, K, V] => F[A]): RedisTransactionB[F, K, V, A] =
+        RedisTransactionB.operation(r)
+      def pure[A](a: A): RedisTransactionB[F, K, V, A] =
+        RedisTransactionB.pure(a)
+    }
+
+    def operation[F[_], K, V, A](r: RedisCommands[F, K, V] => F[A]): RedisTransactionB[F, K, V, A] =
+      Operation(r)
+
+    def pure[F[_], K, V, A](a: A): RedisTransactionB[F, K, V, A] =
+      Pure(a)
+
+    def ap[F[_], K, V, A, B](
+        ff: RedisTransactionB[F, K, V, A => B],
+        fa: RedisTransactionB[F, K, V, A]
+    ): RedisTransactionB[F, K, V, B] =
+      Ap(
+        ff.asInstanceOf[RedisTransactionInternal[F, K, V, A => B]], // Safe ONLY because we are PRIVATE
+        fa.asInstanceOf[RedisTransactionInternal[F, K, V, A]]
+      )
+
+    private sealed trait RedisTransactionInternal[F[_], K, V, A] extends RedisTransactionB[F, K, V, A] {
+      final def transact(cmd: RedisCommands[F, K, V])(implicit F: Concurrent[F], L: Log[F]): F[A] =
+        (Resource.makeCase(cmd.multi) {
+          case (_, ExitCase.Completed) => F.unit
+          case (_, ExitCase.Error(e)) =>
+            cmd.discard *> L.error(s"Transaction failed: ${e.getMessage}")
+          case (_, ExitCase.Canceled) =>
+            cmd.discard *> L.error("Transaction canceled")
+        } >> Resource.liftF(L.info("Transaction started")) >> runInternal(cmd)).use { fib =>
+          cmd.exec *>
```

Comment:
We currently make zero guarantees these are run in transaction as we have not insured the operations have actually been sheduled. Only submitted to the scheduler.

A lot of this could be improved with some sort of locking mechanism as I've mentioned elsewhere. But this is  port of current behavior while keeping type information.